pkg/tfgen/generate_schema.go

```diff
@@ -97,39 +99,68 @@ func gatherSchemaNestedTypesForMember(member moduleMember) map[string]*schemaNes
 func (nt *schemaNestedTypes) gatherFromMember(member moduleMember) {
 	switch member := member.(type) {
 	case *resourceType:
-		nt.gatherFromProperties(member, member.name, "", member.inprops, true)
-		nt.gatherFromProperties(member, member.name, "", member.outprops, true)
+		nt.gatherFromProperties(member, member.name, member.inprops, true, false, true)
+		nt.gatherFromProperties(member, member.name, member.outprops, false, false, true)
 		if !member.IsProvider() {
-			nt.gatherFromProperties(member, member.name, "", member.statet.properties, true)
+			nt.gatherFromProperties(member, member.name, member.statet.properties, false, true, true)
 		}
 	case *resourceFunc:
-		nt.gatherFromProperties(member, member.name, "", member.args, true)
-		nt.gatherFromProperties(member, member.name, "", member.rets, true)
+		nt.gatherFromProperties(member, member.name, member.args, true, false, true)
+		nt.gatherFromProperties(member, member.name, member.rets, false, false, true)
 	case *variable:
-		nt.gatherFromPropertyType(member, member.name, "", "", member.typ, true)
+		nt.gatherFromPropertyType(member, member.name, "", member.typ, false, false, true)
 	}
 }
 
-func (nt *schemaNestedTypes) declareType(
-	declarer declarer, namePrefix, name, nameSuffix string, typ *propertyType, pyMapCase bool) string {
+func (nt *schemaNestedTypes) declareType(declarer declarer, namePrefix, name string, typ *propertyType,
+	isInput, isState, pyMapCase bool) string {
 
 	// Generate a name for this nested type.
-	baseName := namePrefix + strings.Title(name)
+	typeName := namePrefix + strings.Title(name)
 
 	// Override the nested type name, if necessary.
 	if typ.nestedType.Name().String() != "" {
-		baseName = typ.nestedType.Name().String()
+		typeName = typ.nestedType.Name().String()
 	}
 
-	typeName := baseName + strings.Title(nameSuffix)
-	typ.name = typeName
+	if existing, ok := nt.nameToType[typeName]; ok {
+		contract.Assert(existing.declarer == declarer)
+
+		if existing.typ.equals(typ) && pyMapCase == existing.pyMapCase {
+			return typeName
+		}
+
+		// Due to the order in which we declare types, we should only see the following sorts of conflicts:
+		// - input + output
+		// - input + state
+		// - output + state
+		contract.Assert(existing.isInput || isState)
+
+		if isState {
+			typeName = typeName + "State"
+			_, collision := nt.nameToType[typeName]
+			contract.Assert(!collision)
+		} else {
+			existing.typ.name = typeName + "Args"
```

Comment:
Note that unless we decide on a different suffix, these changes will break Go codegen. The Go code generator uses `Args` as the suffix for its generated inputty types, and is guaranteed to cause collisions in these cases.

Reply:
We can either pick a different suffix (e.g. `Inputs`, `Arg`, etc.) or fix the Go code generator.

Reply:
This change also breaks the C# code: the type I get is now called `Inputs.AccountIdentityArgsArgs` (and `Outputs.AccountIdentity` which is the same). Do we need to change the C# codegen now?